Mathlib/Data/Matroid/Basic.lean

```diff
@@ -0,0 +1,884 @@
+  This works fairly well, but there is room for improvement.
+
+  A related decision is to not have matroids themselves be a typeclass.
+  This would make things be notationally simpler
+  (having `Base` in the presence of `[Matroid α]` rather than `M.Base` for a term `M : Matroid α`)
+  but is again just too awkward when one has multiple matroids on the same type.
+  In fact, in regular written mathematics,
+  it is normal to explicitly indicate which matroid something is happening in,
+  so our notation mirrors common practice.
+
+
+## References
+
+[1] The standard text on matroid theory
+[J. G. Oxley, Matroid Theory, Oxford University Press, New York, 2011.]
+
+[2] The robust axiomatic definition of infinite matroids
+[H. Bruhn, R. Diestel, M. Kriesell, R. Pendavingh, P. Wollan, Axioms for infinite matroids,
+  Adv. Math 239 (2013), 18-46]
+
+[3] Equicardinality of matroid bases is independent of ZFC.
+[N. Bowler, S. Geschke, Self-dual uniform matroids on infinite sets,
+  Proc. Amer. Math. Soc. 144 (2016), 459-471]
+-/
+
+
+
+open Set
+
+/-- A predicate `P` on sets satisfies the exchange property if, for all `X` and `Y` satisfying `P`
+  and all `a ∈ X \ Y`, there exists `b ∈ Y \ X` so that swapping `a` for `b` in `X` maintains `P`.-/
+def Matroid.ExchangeProperty {α : Type _} (P : Set α → Prop) : Prop :=
+  ∀ X Y, P X → P Y → ∀ a ∈ X \ Y, ∃ b ∈ Y \ X, P (insert b (X \ {a}))
+
+/-- A set `X` has the maximal subset property for a predicate `P` if every subset of `X` satisfying
+  `P` is contained in a maximal subset of `X` satisfying `P`.  -/
+def Matroid.ExistsMaximalSubsetProperty {α : Type _} (P : Set α → Prop) (X : Set α) : Prop :=
+  ∀ I, P I → I ⊆ X → (maximals (· ⊆ ·) {Y | P Y ∧ I ⊆ Y ∧ Y ⊆ X}).Nonempty
+
+/-- A `Matroid α` is a `ground` set of type `Set α`, and a nonempty collection of its subsets
+  satisfying the exchange property and the maximal subset property. Each such set is called a
+  `Base` of `M`. -/
+@[ext] structure Matroid (α : Type _) where
+  /-- `M` has a ground set. -/
+  (ground : Set α)
+  /-- `M` has a predicate `Base` definining its bases -/
+  (Base : Set α → Prop)
+  /-- There is at least one `Base` -/
+  (exists_base' : ∃ B, Base B)
+  /-- For any bases `B,B'` and `e ∈ B \ B'`, there is some `f ∈ B' \ B` for which `B-e+f`
+    is a base -/
+  (base_exchange' : Matroid.ExchangeProperty Base)
+  /-- Every subset `I` of a set `X` for which `I` is contained in a base is contained in a maximal
+    subset of `X` that is contained in a base. -/
+  (maximality' : ∀ X, X ⊆ ground → Matroid.ExistsMaximalSubsetProperty (∃ B, Base B ∧ · ⊆ B) X)
+  /-- every base is contained in the ground set -/
+  (subset_ground' : ∀ B, Base B → B ⊆ ground)
+
+namespace Matroid
+
+variable {α : Type _} {M : Matroid α}
+
+attribute [pp_dot] Base
+
+/-- We write `M.E` for the ground set of a matroid `M`-/
+@[pp_dot] def E (M : Matroid α) : Set α := M.ground
+
+@[simp] theorem ground_eq_E (M : Matroid α) : M.ground = M.E := rfl
```

Comment:
What do you think of using `E` everywhere, and dropping `.ground` altogether? What about the names? Atm, it seems that statements below use `ground` in the name, but `E` in the actual Lean expression.

Reply:
I would be fine with having `E` as the name of the structure field. But it feels a bit wrong to use `E` in lemma names; one talks about the 'ground set' of a matroid M, writing E(M). It's not referrred to as an 'E-set' or anything like that. IMO, using `M.ground` in place of `M.E` in calculations would add a huge number of characters to everything and make simple things look more complicated. If things were to change, I would prefer the change of replacing 'ground' with 'E' everywhere. How strongly do you feel about this?

Reply:
I think your arguments make sense. In that case, I suggest using `E` everywhere, even in the definition of matroids, but using `ground` in names. Your arguments should also end up in the documentation.